### ISSUE TITLE
bug: missing multiple scalar function definitions for same scalar func with different signatures

### ISSUE THREAD

**Issue by gforsyth** [ibis-project/ibis-substrait#765]
### What happened?

```python
import ibis
from ibis_substrait.compiler.core import SubstraitCompiler


compiler = SubstraitCompiler()


t = ibis.table({'a': 'int64', 'b': 'float32'})
expr = t.mutate(c=t.a + t.a, d=t.b+t.b)
compiler.compile(expr)
```

This should return a Substrait plan with two scalar function entries at the top, with names `add:i64_i64` and `add:fp32_fp32`, respectively, but it only returns the first:

```
version {
  minor_number: 30
  producer: "ibis-substrait"
}
extension_uris {
  extension_uri_anchor: 1
  uri: "https://github.com/substrait-io/substrait/blob/main/extensions/functions_arithmetic.yaml"
}
extensions {
  extension_function {
    extension_uri_reference: 1
    function_anchor: 1
    name: "add:i64_i64"
  }
}
<snip>
```

### What version of ibis-substrait are you using?

main

### What substrait consumer(s) are you using, if any?

n/a

**Comment by gforsyth** [comment]
There's a related bug here, where `compiler` is holding on to previously defined scalar functions and isn't including the type information in the lookup of those scalar functions, so if you only add two `float32` columns you'll get `add:fp32_fp32` for the signature, but if you FIRST add and compile two `int64` columns, then all of the add signatures will be `add:i64_i64`

**Comment by gforsyth** [comment]
fixed in #768
